zebrad/src/components/sync.rs

```diff
@@ -789,34 +834,92 @@ where
         Ok(())
     }
 
-    /// Queue download and verify tasks for each block that isn't currently known to our node
+    /// Queue download and verify tasks for each block that isn't currently known to our node.
+    ///
+    /// TODO: turn obtain and extend tips into a separate task, which sends hashes via a channel?
     async fn request_blocks(
         &mut self,
-        hashes: IndexSet<block::Hash>,
-    ) -> Result<(), BlockDownloadVerifyError> {
-        debug!(hashes.len = hashes.len(), "requesting blocks");
+        mut hashes: IndexSet<block::Hash>,
+    ) -> Result<IndexSet<block::Hash>, BlockDownloadVerifyError> {
+        let lookahead_limit = self.lookahead_limit(hashes.len());
+
+        debug!(
+            hashes.len = hashes.len(),
+            ?lookahead_limit,
+            "requesting blocks",
+        );
+
+        let extra_hashes = if hashes.len() > lookahead_limit {
+            hashes.split_off(lookahead_limit)
+        } else {
+            IndexSet::new()
+        };
+
         for hash in hashes.into_iter() {
             self.downloads.download_and_verify(hash).await?;
         }
 
-        Ok(())
+        Ok(extra_hashes)
+    }
+
+    /// The configured lookahead limit, based on the currently verified height,
+    /// and the number of hashes we haven't queued yet..
+    fn lookahead_limit(&self, new_hashes: usize) -> usize {
```

Comment:
Dynamically calculate different lookahead limits for full and checkpoint verification